### PR TITLE
Improve the performance of connection filters

### LINKED ISSUE #768: StreamSocketOutput.WriteAsync is Sync on Full framework (SSL)
From #569; PR #588 resolves coreclr, however full framework still has the issue.

Issue on full framework if made sync (it blocks)

Fullframework uses framework's SSL stream the corefx SSL steam only supports 

net46 (.NETFramework,Version=v4.6)
netstandard1.3 (.NETStandard,Version=v1.3)

So would currently need to version bump from v4.5.1 to use it or get the sslstream retargeted (if that would even work?)

### PR DESCRIPTION
**Pull request by halter73**
Fixes #768

### CODE REVIEW

**Comment by halter73** [comment]
> dotnet run -- -n plaintext -s http://adx-3u-perfsvr:5001 -c http://adx-3u-svr02:5002 -o  "KestrelHttpServer@halter73/noop-filter-old" -o "Benchmarks@halter73/noop-filter"
    Project Repository (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project Benchmarks.ClientJob (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project Benchmarks.ServerJob (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project BenchmarksDriver (.NETCoreApp,Version=v1.0) was previously compiled. Skipping compilation.
    [05:01:53.027] Starting scenario Plaintext on benchmark server...
    [05:02:19.695] Warmup
    [05:02:19.704] Starting scenario Plaintext on benchmark client...
    [05:02:30.772] Scenario Plaintext completed on benchmark client
    [05:02:30.772] RPS: 890318.48
    [05:02:30.772] Stopping scenario Plaintext on benchmark client...
    [05:02:30.774] Benchmark
    [05:02:30.774] Starting scenario Plaintext on benchmark client...
    [05:02:41.812] Scenario Plaintext completed on benchmark client
    [05:02:41.812] RPS: 927072.76
    [05:02:41.812] Stopping scenario Plaintext on benchmark client...
    [05:02:41.814] Stopping scenario Plaintext on benchmark server...
    > dotnet run -- -n plaintext -s http://adx-3u-perfsvr:5001 -c http://adx-3u-svr02:5002 -o  "KestrelHttpServer@halter73/noop-filter" -o "Benchmarks@halter73/noop-filter"
    Project Repository (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project Benchmarks.ClientJob (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project Benchmarks.ServerJob (.NETStandard,Version=v1.0) was previously compiled. Skipping compilation.
    Project BenchmarksDriver (.NETCoreApp,Version=v1.0) was previously compiled. Skipping compilation.
    [05:04:05.286] Starting scenario Plaintext on benchmark server...
    [05:04:26.974] Warmup
    [05:04:26.987] Starting scenario Plaintext on benchmark client...
    [05:04:38.139] Scenario Plaintext completed on benchmark client
    [05:04:38.139] RPS: 915294.73
    [05:04:38.139] Stopping scenario Plaintext on benchmark client...
    [05:04:38.141] Benchmark
    [05:04:38.141] Starting scenario Plaintext on benchmark client...
    [05:04:49.178] Scenario Plaintext completed on benchmark client
    [05:04:49.178] RPS: 977263.77
    [05:04:49.178] Stopping scenario Plaintext on benchmark client...
    [05:04:49.180] Stopping scenario Plaintext on benchmark server...

Not a super drastic improvement. It looks like wrapping `SocketOutput` is having a *much* larger impact on perf than wrapping `SocketInput`. I haven't yet come up with a great way to improve `StreamSocketOutput` yet.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs [hunk 1]
```diff
@@ -176,9 +170,6 @@ private void ApplyConnectionFilter()
 
             _frame.PrepareRequest = _filterContext.PrepareRequest;
 
-            // Reset needs to be called here so prepare request gets applied
-            _frame.Reset();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs]: ? What changed?

File: test/Microsoft.AspNetCore.Server.KestrelTests/TestHelpers/SocketInputExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Http;
+
+namespace Microsoft.AspNetCore.Server.KestrelTests.TestHelpers
+{
+    public static class SocketInputExtensions
+    {
+        public static void IncomingData(this SocketInput input, byte[] buffer, int offset, int count)
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Server.KestrelTests/TestHelpers/SocketInputExtensions.cs]: So lazy...

File: test/Microsoft.AspNetCore.Server.KestrelTests/TestHelpers/SocketInputExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            var bufferIndex = offset;
+            var remaining = count;
+
+            while (remaining > 0)
+            {
+                var block = input.IncomingStart();
+
+                var bytesLeftInBlock = block.Data.Offset + block.Data.Count - block.End;
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Server.KestrelTests/TestHelpers/SocketInputExtensions.cs]: Wrap the block in an iterator and use CopyFrom

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs [hunk 4]
```diff
@@ -67,39 +67,6 @@ public MemoryPoolBlock IncomingStart()
             return _pinned;
         }
 
-        public void IncomingData(byte[] buffer, int offset, int count)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketInput.cs]: 👏

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameRequestStream.cs [hunk 5]
```diff
@@ -76,8 +76,6 @@ public override int Read(byte[] buffer, int offset, int count)
 #if NET451
         public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
         {
-            ValidateState(default(CancellationToken));
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameRequestStream.cs]: Why did you remove this?

File: src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/FilteredStreamAdapter.cs [hunk 6]
```diff
@@ -30,72 +24,39 @@ public class FilteredStreamAdapter : IDisposable
...
-            try
-            {
-                SocketInput.IncomingFin();
-            }
-            catch (Exception ex)
+            do
             {
-                _log.LogError(0, ex, "FilteredStreamAdapter.OnStreamClose");
-            }
+                var block = SocketInput.IncomingStart();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/FilteredStreamAdapter.cs]: The abort awaiting calls have all disappeared. Since all of the tests pass I'm assuming none of that mattered or we just have missing tests.

File: src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LibuvStream.cs [hunk 7]
```diff
@@ -132,5 +132,83 @@ private ValueTask<int> ReadAsync(ArraySegment<byte> buffer)
         {
             return _input.ReadAsync(buffer.Array, buffer.Offset, buffer.Count);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LibuvStream.cs]: Couldn't you rename this Stream the same time 😄

File: src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/StreamSocketOutput.cs [hunk 8]
```diff
@@ -24,8 +24,6 @@ public class StreamSocketOutput : ISocketOutput
 
         private bool _canWrite = true;
 
-        private object _writeLock = new object();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/StreamSocketOutput.cs]: But But, overlapping async writes will get corrupted 😄

File: src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/FilteredStreamAdapter.cs (hunk 6 above)
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/FilteredStreamAdapter.cs]: They did matter when we used to call `SocketInput.IncomingFin();` in `OnSocketClosed()`. We no longer do, so we can simplify.

File: src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LibuvStream.cs (hunk 7 above)
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LibuvStream.cs]: `ChannelStream`? I could also move `SocketInput` out of the `Http` folder, but didn't feel it was related to this change. Maybe it's worth moving everything around and cleaning up as part of this PR.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameRequestStream.cs (hunk 5 above)
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/FrameRequestStream.cs]: It was wrong. This wasn't even looking at the task returned from `Validate` state so was somewhat broken. `ReadAsync` calls `ValidateState`.

File: test/Microsoft.AspNetCore.Server.KestrelTests/TestHelpers/SocketInputExtensions.cs (hunk 3 above)
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/TestHelpers/SocketInputExtensions.cs]: That would require a change to `IncomingComplete`. Right now, `IncomingStart/Complete` expects one block at a time.

Comment (davidfowl) [inline test/Microsoft.AspNetCore.Server.KestrelTests/TestHelpers/SocketInputExtensions.cs]: We'll fix this in the next one. Thats how it should work 😄

File: src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LibuvStream.cs (hunk 7 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Filter/Internal/LibuvStream.cs]: Sure, we can do it another PR

**Comment by davidfowl** [comment]
:shipit:

**Comment by mikeharder** [comment]
@halter73: Please wait until I've added this scenario to Benchmarks (https://github.com/aspnet/benchmarks/pull/109) so we can see the perf improvement in our automated results

**Comment by mikeharder** [comment]
It's hard to tell since the results have a lot of variance, but I'm not seeing a clear improvement from this change:

```
Before  After
------  -----
876     935
964     847
860     942
919     891
855     910
878     862

Mean(Before): 892
Mean(After): 896

Before: dotnet run -- -n plaintext --connectionFilter Benchmarks.PassthroughConnectionFilter -o KestrelHttpServer@2244f19
After: dotnet run -- -n plaintext --connectionFilter Benchmarks.PassthroughConnectionFilter -o KestrelHttpServer@halter73/better-filters
```

**Comment by benaadams** [comment]
Does it fix the ssl async-over-sync on full framework as suggested? 

If so I'd take it regardless...

**Comment by mikeharder** [comment]
@benaadams: How would one test the "ssl async-over-sync"?  Just benchmark plaintext over HTTPS using full framework rather than core?

**Comment by benaadams** [comment]
Hmm... might be better if I email you the exact pain points to push rather than post here?

**Comment by benaadams** [comment]
@mikeharder try plaintext over HTTPS on full framework, but switch the `Invoke` for 
```csharp
public async Task Invoke(HttpContext httpContext)
{
    var response = httpContext.Response;
    response.StatusCode = 200;
    response.ContentType = "text/plain";
    response.Headers["Content-Length"] = "13";

    for (var i = 0; i < _helloWorldPayload.Length; i++)
    {
        await response.Body.WriteAsync(_helloWorldPayload, i, 1);
    }
}
```
Or probably better
```csharp
private static readonly byte[] _payload = Encoding.UTF8.GetBytes(new string(' ', 65537));
public async Task Invoke(HttpContext httpContext)
{
    var response = httpContext.Response;
    response.StatusCode = 200;
    response.ContentType = "text/plain";

    for (var i = 0; i < 12; i++)
    {
        await response.Body.WriteAsync(_payload, 0, _payload.Length);
    }
}
```
Then crank up the connection count...

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 9]
```diff
@@ -312,6 +311,7 @@ public void Reset()
         /// </summary>
         public void Start()
         {
+            Reset();
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Why did you move this here? It was bad in the ctor so I like that you moved it here, but just wondering if there were other reasons.

**Comment by cesarbs** [comment]
lgtm 🚢  (but wait for @mikeharder 😄 )

**Comment by davidfowl** [comment]
Regardless of the performance improvement, the code is better and fixes some problems (async await FTW)

**Comment by mikeharder** [comment]
I'm not opposed to the change, just wanted to make it clear I don't see a perf benefit in Plaintext.  I'll test JSON as well.

**Comment by benaadams** [comment]
Not seeing any difference with https on full framework with the above middleware (pre and post)?

**Comment by mikeharder** [comment]
@benaadams Haven't tested full framework yet, will try today.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs (hunk 9 above)
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Mostly so we don't have to call it in `Connection.ApplyConnectionFilter()`.

**Comment by halter73** [comment]
The benchmarks show that SSL on full framework is much better (read less broken) after this change. The BenchmarksDriver doesn't support full framework atm, so I ran some tests on my dev machine. This was enough to show a big difference. Basically after the change, Kestrel doesn't fall over with a bunch of HTTS connections on full framework:

    [shalter@shalter-arch wrk]$ ./wrk -c 64 -t 8 -d 10 -s pipeline.lua https://winhost:5001 -- 16
    Running 10s test @ https://winhost:5001
      8 threads and 64 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency     0.00us    0.00us   0.00us    -nan%
        Req/Sec     0.00      0.00     0.00      -nan%
      0 requests in 10.10s, 0.00B read
    Requests/sec:      0.00
    Transfer/sec:       0.00B
    [shalter@shalter-arch wrk]$ ./wrk -c 16 -t 8 -d 10 -s pipeline.lua https://winhost:5001 -- 16
    Running 10s test @ https://winhost:5001
      8 threads and 16 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    13.31ms   26.70ms 386.70ms   92.22%
        Req/Sec     2.03k   551.20     4.10k    70.57%
      128749 requests in 10.07s, 16.33MB read
    Requests/sec:  12782.50
    Transfer/sec:      1.62MB
    [shalter@shalter-arch wrk]$ ./wrk -c 64 -t 8 -d 10 -s pipeline.lua https://winhost:5001 -- 16
    Running 10s test @ https://winhost:5001
      8 threads and 64 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency     0.00us    0.00us   0.00us    -nan%
        Req/Sec     0.00      0.00     0.00      -nan%
      0 requests in 10.10s, 0.00B read
    Requests/sec:      0.00
    Transfer/sec:       0.00B

#### halter73/better-filters below

    [shalter@shalter-arch wrk]$ ./wrk -c 64 -t 8 -d 10 -s pipeline.lua https://winhost:5001 -- 16
    Running 10s test @ https://winhost:5001
      8 threads and 64 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    33.01ms   39.10ms 469.31ms   92.12%
        Req/Sec     2.76k   547.24     6.31k    79.97%
      215326 requests in 10.05s, 27.31MB read
    Requests/sec:  21417.80
    Transfer/sec:      2.72MB
    [shalter@shalter-arch wrk]$ ./wrk -c 256 -t 8 -d 10 -s pipeline.lua https://winhost:5001 -- 16
    Running 10s test @ https://winhost:5001
      8 threads and 256 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency    99.78ms   60.76ms   1.11s    47.64%
        Req/Sec     2.55k   828.66     5.83k    74.93%
      197129 requests in 10.08s, 25.00MB read
      Socket errors: connect 0, read 0, write 0, timeout 12
    Requests/sec:  19549.64
    Transfer/sec:      2.48MB
    [shalter@shalter-arch wrk]$ ./wrk -c 1024 -t 8 -d 10 -s pipeline.lua https://winhost:5001 -- 16
    Running 10s test @ https://winhost:5001
      8 threads and 1024 connections
      Thread Stats   Avg      Stdev     Max   +/- Stdev
        Latency   387.93ms  233.96ms   1.98s    28.72%
        Req/Sec     2.24k     1.58k    7.27k    69.36%
      150779 requests in 10.08s, 19.12MB read
      Socket errors: connect 11, read 0, write 0, timeout 12
    Requests/sec:  14956.46
    Transfer/sec:      1.90MB

**Comment by benaadams** [comment]
I thought it might make a difference in a big way; have removed the example code.

:rocket:
